mitodl/ocw-studio
# feat: update drive sync and allow file deletion

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1555
Relevant https://github.com/mitodl/hq/discussions/359

#### What's this PR do?

This PR:
- Updates the website GDrive sync task
- Deletes files that no longer exist in GDrive and aren't being used on any page
  - It was necessary to delete existing files to achieve the effect expected in the linked ticket. Since I was writing the delete logic, I also considered https://github.com/mitodl/hq/discussions/359 in the implementation.
  - Note: Files that are used in content other than `page` will also be deleted.
- The rest of the details can be read from the issue comments.

#### Any background context you want to provide?

Might be irrelevant but in case you wonder, I experimented with the [changes API](https://developers.google.com/drive/api/v3/reference/changes) and decided not to use it for two reasons.
1. The API does not provide any filtering options.
2. A Change can be removed from the list because of a user action. The documentation does not specify a strict set of reasons.
3. In the end, even if possible, it would've been more complicated to implement.

#### How should this be manually tested?
1. Download the folder [here](https://drive.google.com/drive/folders/1fJ0Sv5ZKy-NKCdfnLEOAcM-jrEZuMU0P?usp=sharing)
2. Head to your local ocw-studio setup
4. Checkout this branch `hussaintaj/1555-gdrive-sync`
5. Start ocw-studio and create a test course.
6. Open the folder you downloaded in step 1. Open the `step01` subdirectory. Upload these files to the associated course's GDrive folder `final_files`.
7. In ocw-studio, open the resources tab of your test course and hit `Sync w/ GDrive`.
8. **Expected Behaviour**: You should see 5 files after a successful sync.
9. Click on `image3.png` and add something in the description field. This will act as a marker that we will use later.
10. Create a page in your test course. In the body, embed `image5.png`. Save the page.
11. Finish adding the metadata for the course and then publish it to draft or live, or both.
12. Open your local instance of MinIO. Log in and navigate to your course files.
13. **Expected Result**: You should see the synced files in the `static_resources`, and content in the `resources` directories.
14. Open the GDrive folder from step 5. Delete the following files `image3.png`, `image4.png`, `image5.png`.
15. Open the folder you downloaded in step 1. Open the `step02` subdirectory.
    * Upload `image2.png`. Gdrive will ask you if you want to replace it. Say yes.
    * Upload `image3.png`.
16. Repeat step 6
17. After sync. **Expected Result**
    * `image4.png` is removed from the resources file.
    * `image5.png` is not removed. Because it was used in step 9.
    * Click `image3.png`, the description should have the content you added in step 8.
18. Go to your GDrive trash. Restore `image3.png` from step 13.
19. Repeat step 6
20. After sync. **Expected Result**: You should see two `image3.png`.
21. Publish the course again.
22. Repeat step 11 and 12. Make sure all the changes in the files are reflected properly. Note: One file will be named `image32.png`
23. That's it for this systematic testing. Feel free to go crazy and try other things that you can think of.
